receiver/prometheusexecreceiver/receiver.go

```diff
@@ -0,0 +1,313 @@
+func (wrapper *prometheusReceiverWrapper) computeHealthAndCrashCount(elapsed time.Duration, crashCount int) (int, error) {
+	// Reset crash count to 1 if the process seems to be healthy now, else increase crashCount
+	if elapsed > healthyProcessTime {
+		return 1, nil
+	}
+	crashCount++
+
+	// Stop the associated receiver until process starts back up again since it is unhealthy (too little elapsed time and high crashCount) or if port is generated, to allow for new port
+	if wrapper.originalPort == 0 || crashCount > healthyCrashCount {
+		err := wrapper.Shutdown(wrapper.context)
+		if err != nil {
+			return crashCount, fmt.Errorf("could not stop receiver associated to process, killing it")
+		}
+	}
+
+	return crashCount, nil
+}
+
+// assignNewRandomPort generates a new port, creates a new metrics receiver with that port and assigns it to the wrapper
+func (wrapper *prometheusReceiverWrapper) assignNewRandomPort(oldPort int) (int, error) {
+	var err error
+	newPort := generateRandomPort(oldPort)
+
+	// Assign the new port in the config
+	wrapper.receiverConfig.PrometheusConfig.ScrapeConfigs[0].ServiceDiscoveryConfig.StaticConfigs[0].Targets = []model.LabelSet{
+		{model.AddressLabel: model.LabelValue(fmt.Sprintf("localhost:%v", newPort))},
+	}
+
+	// Create new Prometheus receiver with new config and replace pointer to it in wrapper
+	factory := &prometheusreceiver.Factory{}
+	wrapper.prometheusReceiver, err = factory.CreateMetricsReceiver(wrapper.context, wrapper.logger, wrapper.receiverConfig, wrapper.consumer)
+	if err != nil {
+		return 0, fmt.Errorf("unable to create Prometheus receiver. Killing this process")
+	}
+
+	return newPort, nil
+}
+
+// fillPortPlaceholders will check if any of the strings in the process data have the {{port}} placeholder, and replace it if necessary
+func (wrapper *prometheusReceiverWrapper) fillPortPlaceholders(newPort int) *subprocessmanager.SubprocessConfig {
+	var port string
+	if wrapper.originalPort == 0 {
+		port = strconv.Itoa(newPort)
+	} else {
+		port = strconv.Itoa(wrapper.originalPort)
+	}
+
+	newConfig := *wrapper.subprocessConfig
+	// ReplaceAll runs much faster (about 5x according to my tests) than checking for a regex match, therefore no checks are made and ReplaceAll simply returns the original string if no match is found
+	// ReplaceAll is run on the original strings of the config, which remain unchanged
+	newConfig.Command = strings.ReplaceAll(wrapper.config.SubprocessConfig.Command, portTemplate, port)
+
+	for i, env := range wrapper.config.SubprocessConfig.Env {
+		newConfig.Env[i].Value = strings.ReplaceAll(env.Value, portTemplate, port)
+	}
+
+	return &newConfig
+}
+
+// generateRandomPort will try to generate a random port until it is different than the last one generated
+func generateRandomPort(lastPort int) int {
+	var newPort int
+	for {
+		newPort = random.Intn(maxPortRange-minPortRange) + minPortRange
+
+		// Generate another port if it's the same
```

Comment:
Because there was concern in the design doc of race conditions, where checking for a port to listen on wasn't enough, and that if another program on the VM stole that port before we could use it (but after we'd checked it), we'd need to generate a new port - since the exporter would not be able to bind to it. This way, a new port is generated at every failure, to mitigate the fact that there are race conditions. In an ideal world, we'd **only** do this if the error returned by the subprocess is an error relating to port being in use. But there's isn't a way to reliably do this unfortunately.